File: src/android/TalkingDataEAuthPlugin.java
package com.talkingdata.analytics;

import android.content.Context;

import com.tendcloud.tenddata.TalkingDataEAuth;
import com.tendcloud.tenddata.TalkingDataEAuthCallback;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class TalkingDataEAuthPlugin extends CordovaPlugin implements TalkingDataEAuthCallback {
	Context ctx;
	CallbackContext callback;
	
	@Override
	public void initialize(CordovaInterface cordova, CordovaWebView webView) {
		super.initialize(cordova, webView);
		this.ctx = cordova.getActivity().getApplicationContext();
	}
	
	@Override
	public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
		if (action.equals("init")) {
			String appId = args.getString(0);
			String secretId = args.getString(1);
			TalkingDataEAuth.initEAuth(ctx, appId, secretId);
			return true;
		} else if (action.equals("applyAuthCode")) {
			String countryCode = args.getString(0);
            String mobile = args.getString(1);
            TalkingDataEAuth.TDAuthCodeType type = this.intToAuthCodeType(args.getInt(2));
            String accountName = args.getString(3);
            callback = callbackContext;
			TalkingDataEAuth.applyAuthCode(countryCode, mobile, this, accountName, type);
			return true;
		} else if (action.equals("reapplyAuthCode")) {
            String countryCode = args.getString(0);
            String mobile = args.getString(1);
            TalkingDataEAuth.TDAuthCodeType type = this.intToAuthCodeType(args.getInt(2));
            String accountName = args.getString(3);
            String requestId = args.getString(4);
            callback = callbackContext;
            TalkingDataEAuth.reapplyAuthCode(countryCode, mobile, requestId, this, accountName, type);
			return true;
		} else if (action.equals("isVerifyAccount")) {
			String accountName = args.getString(0);
			callback = callbackContext;
			TalkingDataEAuth.isVerifyAccount(accountName, this);
			return true;
    	} else if (action.equals("isMobileMatchAccount")) {
    		String accountName = args.getString(0);
    		String countryCode = args.getString(1);
    		String mobile = args.getString(2);
    		callback = callbackContext;
    		TalkingDataEAuth.isMobileMatchAccount(countryCode, mobile, accountName, this);
    		return true;
    	} else if (action.equals("bind")) {
    		String countryCode = args.getString(0);
    		String mobile = args.getString(1);
    		String authCode = args.getString(2);
    		String accountName = args.getString(3);
    		callback = callbackContext;
    		TalkingDataEAuth.bindEAuth(countryCode, mobile, authCode, this, accountName);
    		return true;
    	} else if (action.equals("unbind")) {
    		String countryCode = args.getString(0);
    		String mobile = args.getString(1);
    		String accountName = args.getString(2);
    		callback = callbackContext;
    		TalkingDataEAuth.unbindEAuth(countryCode, mobile, accountName, this);
			return true;
		}
		return false;
	}

    private TalkingDataEAuth.TDAuthCodeType intToAuthCodeType(final int type) {
        if (type == 1) {
            return TalkingDataEAuth.TDAuthCodeType.voiceCallAuth;
        } else {
            return TalkingDataEAuth.TDAuthCodeType.smsAuth;
        }
    }


	@Override
	public void onRequestSuccess(TalkingDataEAuth.TDEAuthType type, String requestId, String phoneNumber, JSONArray phoneNumSeg) {
		try {
			List<PluginResult> params = new ArrayList<>();
			params.add(new PluginResult(PluginResult.Status.OK, type.toString()));
			params.add(new PluginResult(PluginResult.Status.OK, requestId));
			params.add(new PluginResult(PluginResult.Status.OK, phoneNumber));
            phoneNumSeg = phoneNumSeg == null ? new JSONArray() : phoneNumSeg;
			params.add(new PluginResult(PluginResult.Status.OK, phoneNumSeg));
			callback.sendPluginResult(new PluginResult(PluginResult.Status.OK, params));
		}catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public void onRequestFailed(TalkingDataEAuth.TDEAuthType type, int errorCode, String errorMessage) {
		try {
			List<PluginResult> params = new ArrayList<>();
			params.add(new PluginResult(PluginResult.Status.ERROR, type.toString()));
			params.add(new PluginResult(PluginResult.Status.ERROR, errorCode));
			params.add(new PluginResult(PluginResult.Status.ERROR, errorMessage));
			callback.sendPluginResult(new PluginResult(PluginResult.Status.ERROR, params));
		}catch (Exception e) {
			e.printStackTrace();
		}
    }
}

File: src/ios/TalkingDataEAuthPlugin.m
//
//  TalkingDataEAuthPlugin.m
//  TalkingData
//
//  Created by liweiqiang on 15-12-2.
//
//

#import "TalkingDataEAuthPlugin.h"
#import "TalkingDataEAuth.h"

@interface TalkingDataEAuthPlugin () <TalkingDataEAuthDelegate>

@property (nonatomic, strong) NSString *callbackId;

@end

@implementation TalkingDataEAuthPlugin

- (void)init:(CDVInvokedUrlCommand*)command {
    NSString *appId = [command.arguments objectAtIndex:0];
    if (![appId isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *secretId = [command.arguments objectAtIndex:1];
    if (![secretId isKindOfClass:[NSString class]]) {
        return;
    }
    [TalkingDataEAuth initEAuth:appId secretId:secretId];
}

- (void)applyAuthCode:(CDVInvokedUrlCommand*)command {
    NSString *countryCode = [command.arguments objectAtIndex:0];
    if (![countryCode isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *mobile = [command.arguments objectAtIndex:1];
    if (![mobile isKindOfClass:[NSString class]]) {
        return;
    }
    NSNumber *typeNum = [command.arguments objectAtIndex:2];
    if (![typeNum isKindOfClass:[NSNumber class]]) {
        return;
    }
    TDAuthCodeType type = [typeNum integerValue];
    NSString *accountName = [command.arguments objectAtIndex:3];
    if (![accountName isKindOfClass:[NSString class]]) {
        return;
    }
    self.callbackId = command.callbackId;
    [TalkingDataEAuth applyAuthCode:countryCode mobile:mobile authCodeType:type accountName:accountName delegate:self];
}

- (void)reapplyAuthCode:(CDVInvokedUrlCommand*)command {
    NSString *countryCode = [command.arguments objectAtIndex:0];
    if (![countryCode isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *mobile = [command.arguments objectAtIndex:1];
    if (![mobile isKindOfClass:[NSString class]]) {
        return;
    }
    NSNumber *typeNum = [command.arguments objectAtIndex:2];
    if (![typeNum isKindOfClass:[NSNumber class]]) {
        return;
    }
    TDAuthCodeType type = [typeNum integerValue];
    NSString *accountName = [command.arguments objectAtIndex:3];
    if (![accountName isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *requestId = [command.arguments objectAtIndex:4];
    if (![requestId isKindOfClass:[NSString class]]) {
        return;
    }
    self.callbackId = command.callbackId;
    [TalkingDataEAuth reapplyAuthCode:countryCode mobile:mobile authCodeType:type accountName:accountName requestId:requestId delegate:self];
}

- (void)isVerifyAccount:(CDVInvokedUrlCommand*)command {
    NSString *accountName = [command.arguments objectAtIndex:0];
    if (![accountName isKindOfClass:[NSString class]]) {
        return;
    }
    self.callbackId = command.callbackId;
    [TalkingDataEAuth isVerifyAccount:accountName delegate:self];
}

- (void)isMobileMatchAccount:(CDVInvokedUrlCommand*)command {
    NSString *accountName = [command.arguments objectAtIndex:0];
    if (![accountName isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *countryCode = [command.arguments objectAtIndex:1];
    if (![countryCode isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *mobile = [command.arguments objectAtIndex:2];
    if (![mobile isKindOfClass:[NSString class]]) {
        return;
    }
    self.callbackId = command.callbackId;
    [TalkingDataEAuth isMobileMatchAccount:accountName countryCode:countryCode mobile:mobile delegate:self];
}

- (void)bind:(CDVInvokedUrlCommand*)command {
    NSString *countryCode = [command.arguments objectAtIndex:0];
    if (![countryCode isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *mobile = [command.arguments objectAtIndex:1];
    if (![mobile isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *authCode = [command.arguments objectAtIndex:2];
    if (![authCode isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *accountName = [command.arguments objectAtIndex:3];
    if (![accountName isKindOfClass:[NSString class]]) {
        return;
    }
    self.callbackId = command.callbackId;
    [TalkingDataEAuth bindEAuth:countryCode mobile:mobile authCode:authCode accountName:accountName delegate:self];
}

- (void)unbind:(CDVInvokedUrlCommand*)command {
    NSString *countryCode = [command.arguments objectAtIndex:0];
    if (![countryCode isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *mobile = [command.arguments objectAtIndex:1];
    if (![mobile isKindOfClass:[NSString class]]) {
        return;
    }
    NSString *accountName = [command.arguments objectAtIndex:2];
    if (![accountName isKindOfClass:[NSString class]]) {
        return;
    }
    self.callbackId = command.callbackId;
    [TalkingDataEAuth unbindEAuth:countryCode mobile:mobile accountName:accountName delegate:self];
}

- (void)onRequestSuccess:(TDEAuthType)type requestId:(NSString *)requestId phoneNumber:(NSString *)phoneNumber phoneNumSeg:(NSArray *)phoneNumSeg {
    NSMutableArray *args = [NSMutableArray array];
    [args addObject:[NSNumber numberWithInteger:type]];
    [args addObject:requestId ?: @""];
    [args addObject:phoneNumber ?: @""];
    [args addObject:phoneNumSeg ?: @[]];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsMultipart:args];
#if CORDOVA_VERSION_MIN_REQUIRED < __CORDOVA_3_6_0
    [self writeJavascript:[pluginResult toSuccessCallbackString:self.callbackId]];
#else
    [self.commandDelegate sendPluginResult:pluginResult callbackId:self.callbackId];
#endif
}

- (void)onRequestFailed:(TDEAuthType)type errorCode:(NSInteger)errorCode errorMessage:(NSString *)errorMessage {
    NSMutableArray *args = [NSMutableArray array];
    [args addObject:[NSNumber numberWithInteger:type]];
    [args addObject:[NSNumber numberWithInteger:errorCode]];
    [args addObject:errorMessage ?: @""];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsMultipart:args];
#if CORDOVA_VERSION_MIN_REQUIRED < __CORDOVA_3_6_0
    [self writeJavascript:[pluginResult toErrorCallbackString:self.callbackId]];
#else
    [self.commandDelegate sendPluginResult:pluginResult callbackId:self.callbackId];
#endif
}

@end
